collect_app/src/test/java/org/odk/collect/android/configure/JsonSchemaSettingsValidatorTest.kt

```diff
@@ -0,0 +1,68 @@
+package org.odk.collect.android.configure
+
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
+import org.junit.Test
+
+class JsonSchemaSettingsValidatorTest {
```

Comment:
Don't you think it would make sense to test our real settings too with their schema?

Reply:
This is a good thing to discuss. The schema validation with real settings gets tested (as part of the full stack) through a few different places already - `SettingsImporterRegressionTest`, `ConfigureWithQRCodeTest` for example. I think if we were starting from scratch it would be nice to have been adding a test for each setting as we went (checking it accepts and rejects the right things) but backfilling this now would be a massive job. 

The weird thing that comes up in my head whenever I think about adding tests for the schema itself is that once we're sharing this with Central the schema itself becomes the source of truth. In that world, our job is just to integrate with it (which the tests in `JsonSchemaSettingsValidatorTest` ensure to some degree). Maybe we adding a case to `SettingsImporterRegressionTest` to check the "all settings" case (testing almost the full stack of the import) or maybe we want to make sure the `enum` settings all work correctly seeing as they're a bit more complex? Do you have an idea of what kind of testing you'd like to see?

Reply:
The thing is that those test we already have verify just a narrow set of prefs. I don't know... Here you are just testing an external library with some simple data I don't think it's important if you took that library it should have it's own tests so I think that it would be more important to test our scheme if it works correctly instead of the library (of course if you do that with our scheme you will test both).

Reply:
> Here you are just testing an external library with some simple data

Yeah `JsonSchemaValidatorTest` is really just checking our implementation integrates with the library correctly - it doesn't have anything to do with our actual schema. I think that's good to have, but I do agree that we might want more. Do you think we should just add a smoke test that checks importing a JSON with all settings? I think it might also be nice to backfill tests for validating the different `enum` values but that would be better as a separate issue I think.

Reply:
Ok so I guess `SettingsImporterTest` could be improved... We should verify cases like:
- Generally invalid jsons should be ignored
- General Settings should be required
- Admin settings should be required
- Unknown keys should be ignored 
- If enums used values other than listed in enums should be ignored

some of those cases are already covered.

> Do you think we should just add a smoke test that checks importing a JSON with all settings?

Yes I think it would make sense to do that to make sure our scheme is not broken but also that importing all the keys works properly.

Reply:
Now as I look at `SettingsImporter` a lot of things happen there. For example we first load all prefs from json and then after doing that we remove those unknown. This is weird. It would be better to sanitize our json first. We can remove objects from json right so we could go through it and remove unknown elements first so that later we can import everything and don't think about some potential issues.

Reply:
@seadowg what do you think about those tests? are you going to add anything?

Reply:
Yup sorry will be back to this soon!